# Lecture 2 - Python Data Types, Statements

[![View notebook on Github](https://img.shields.io/static/v1.svg?logo=github&label=Repo&message=View%20On%20Github&color=lightgrey)](https://github.com/avakanski/Fall-2026-Applied-Data-Science-with-Python/blob/main/Lectures/Theme_1-Python_Programming/Lecture_2-Python_Data_Types%2C_Statements/Lecture_2-Python_Data_Types%2C_Statements.ipynb)
[![Open In Collab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/avakanski/Fall-2026-Applied-Data-Science-with-Python/blob/main/Lectures/Theme_1-Python_Programming/Lecture_2-Python_Data_Types%2C_Statements/Lecture_2-Python_Data_Types%2C_Statements.ipynb) 

<a id='top'></a>

- [2.1 Python Data Types](#2.1-python-data-types)
    - [2.1.1 Introduction](#2.1.1-introduction)
    - [2.1.2 Numbers](#2.1.2-numbers)
    - [2.1.3 Strings](#2.1.3-strings)
    - [2.1.4 Lists](#2.1.4-lists)
    - [2.1.5 Dictionaries](#2.1.5-dictionaries)
    - [2.1.6 Tuples](#2.1.6-tuples)
    - [2.1.7 Sets](#2.1.7-sets)
    - [2.1.8 Other Data Types](#2.1.8-other-data-types)
    - [2.1.9 String Formatting](#2.1.9-string-formatting)
- [2.2 Statements](#2.2-statements)
    - [2.2.1 if, else, elif Statements](#2.2.1-if,-else,-elif-statements)
    - [2.2.2 for Loops](#2.2.2-for-loops)
    - [2.2.3 while Loops](#2.2.3-while-loops)
    - [2.2.4 break, continue, pass Statements](#2.2.4-break,-continue,-pass-statements)
- [Appendix: Python Interpreter](#appendix:-python-interpreter)
- [References](#references)

<a id="2.1-python-data-types"></a>

## 2.1 Python Data Types

The figure below lists the main data types in Python, and provides information about the category and mutability of the data types.

<img src="images/pic1.jpg" width="350">
<em>Figure source: Reference [1].</em>

<a id="2.1.1-introduction"></a>

### 2.1.1 Introduction

Python is an interpreted, high-level, general-purpose programming language. Created by Guido van Rossum and first released in 1991, Python's design philosophy emphasizes code readability. Its language constructs and object-oriented approach aim to help programmers write clear, logical code for small-scale and large-scale projects. Python interpreters are available for many operating systems. In recent years, Python has become the primary language for Machine Learning and Data Science applications. 

The Python Software Foundation (PSF) is a non-profit organization that manages and directs resources for Python development. Python 3.0 was released in 2008, and it was a significant revision of the language that is not entirely backward-compatible, and much Python 2 code does not run unmodified on Python 3. This course makes use of Python 3. 

#### Dynamic Typing

Python uses *dynamic typing*, meaning that we don't need to declare the type of variable when writing the code, and the type is determined at runtime based on the assigned value. Therefore, we can reassign different data types to variables. This makes Python very flexible in assigning data types, and it differs from other languages that are *statically typed* (such as Java, C, C++) and require to declare the type of a variable before using it, and the type is fixed after it is declared.

In [1]:
# Assign the number 2 to the variable 'my_dogs'
my_dogs = 2

In [2]:
# Show the variable 'my_dogs'
my_dogs

2

In [3]:
# Reassign the list to the variable 'my_dogs' 
my_dogs = ['Sammy', 'Frankie']

In [4]:
# Show the variable 'my_dogs'
my_dogs

['Sammy', 'Frankie']

In the above example, the number `2` was first assigned to the variable `my_dogs`, and afterward the list `['Sammy', 'Frankie']` was assigned to the variable `my_dogs`.

When we create a variable name in Python, we reserve a memory location to store an object. E.g., the variable name `my_dogs` first acts as a reference to the memory location which holds the number `2`. Or, we can think of the variable as a pointer to the memory location where the value `2` is stored. Whenever we use the variable `my_dogs` in our code, Python will retrieve the value `2` from the memory location and associate it with the name `my_dogs`. By assigning the list `['Sammy', 'Frankie']` to the variable `my_dogs`, we instruct Python to associate the name `my_dogs` with a new memory location where the list `['Sammy', 'Frankie']` is stored.

**Pros of Dynamic Typing**

* Very easy to work with
* Faster development time

**Cons of Dynamic Typing**

* May result in unexpected bugs due to type-related errors, which can lead to hideous debugging sessions
* Requires to be aware of the type of objects
* Statically typed languages are more efficient because the compiler can optimize the code based on knowledge of the type of variables

#### Assigning Objects to Variables

When assigning objects to variables in Python, we need to obey the following rules for the names of variables.

* Names can not start with a number
* Names can not contain spaces, use `_` (underscore) instead
* Names can not contain any of these symbols  `:'",<>/?|\!@#%^&*~-+`
* It's considered best practice (according to [PEP8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names)) that names are written with lowercase letters with underscores
* Avoid using Python built-in keywords like `list` and `str` in variable names
* Avoid using the single characters `l` (lowercase letter L), `O` (uppercase letter O) and `I` (uppercase letter I), since they can be confused with `1` and `0`

Variable assignment has the syntax `name = object`, where a single equal sign `=` is used as an **assignment operator**.

In [5]:
# Assign the integer object 5 to the variable name 'a' 
a = 5

In [6]:
# Show the variable 'a'
a

5

As we mentioned, dynamic typing in Python allows variables to be reassigned.

In [7]:
a = 10
a

10

Python also allows to reassign a variable with a reference to the same object. 

In [8]:
# Add 15 to the current value of 'a' and assign it to 'a'
a = a + 15
a

25

Python allows using shortcuts to add, subtract, multiply, and divide numbers with re-assignment using `+=`, `-=`, `*=`, and `/=`.

For instance, `a += 10` is equivalent to `a = a + 10`

In [9]:
# a = a + 10
a += 10
a

35

In [10]:
# a = a * 2
a *= 2
a

70

#### Determining Variable Type with `type()`

Python offers several **built-in functions**, which can perform actions on objects. For instance, we can check the type of the object that is assigned to a variable using Python's built-in function `type()`. 

In [11]:
type(a)

int

In the above example, the type of the variable `a` is *integer*.

In [12]:
# In Python we create tuples with parentheses 
a = (1,2)

In [13]:
type(a)

tuple

It is also important to note that the double equal sign operator `==` in Python is used to test the equality of two expressions, whereas the single equal sign operator `=` is used to assign objects to variables. 

Also, testing for inequality is performed with the not equal operator `!=` . 

The operators greater than `>`, less than `<`, greater than or equal `>=`, less than or equal `<=` perform as would generally be expected. 

<a id="2.1.2-numbers"></a>

### 2.1.2 Numbers

###### Python has several various types of number objects. 

**Integers** are whole numbers, and can be positive or negative. For example: 2 and -2.

**Floating point** numbers in Python have a decimal point, or use an exponential (E) to define the number. For example, 2.0 and -2.163 are examples of floating point numbers. `4E2` (4 times 10 to the power of 2 = 400) and `1E-3` (3 times 10 to the power of -3 = 0.001) are also examples of floating point numbers in Python.

Other types of number objects that are less frequently used include:

* **Complex numbers**, have real and imaginary parts, e.g., 3+4j
* **Decimal numbers**, have control over the precision and rounding of numbers, e.g., `Decimal('0.1')` (see examples in the next section)
*  **Fractions**, are rational numbers with numerator and denominator, e.g., `Fraction(1,3)` = 1/3.

#### Basic Arithmetic Operations

In [14]:
# Addition
2+1

3

In [15]:
# Subtraction
2-1

1

In [16]:
# Multiplication
2*2

4

In [17]:
# Division
3/2

1.5

In [18]:
# Floor Division
7//4

1

The floor division operator `//` (two forward slashes) truncates the decimal number without rounding, and returns an integer result.

If we just want the remainder after division, we use the `%` modulo operator.

In [19]:
# Modulo (remainder)
7%4

3

In [20]:
# Power (exponentiation)
2**3

8

In [21]:
# Can also do roots (e.g., square root is **0.5)
4**0.5

2.0

In [22]:
# Order of operations followed in Python
# Precedence: Parenthesis, Exponentiation, Division, Multiplication, Addition, Subtraction
# E.g., multiplication has precedence over addition
2 + 10 * 10 + 3

105

In [23]:
# Can use parentheses to specify orders
(2+10) * (10+3)

156

Note that floating-point numbers are implemented in computer hardware as binary fractions (fractions of 0 and 1). As a result, many decimal fractions cannot be accurately represented as binary fractions. For example, the decimal number 0.1 results in an infinitely long binary fraction of 0.000110011001100110011.... Since our computer can only store a finite number of decimal places, this will only approximate the above binary fraction, but the approximation will not be equal to 0.1. Hence, such approximations of decimal numbers is the limitation of our computer hardware and not an error in Python.

In [24]:
# Note that display issue in Python, due to using binary fractions to represent float numbers
f3 = 0.1 + 0.1 + 0.1 
f3

0.30000000000000004

In [25]:
# One solution to that is to use decimal numbers, since they have rounding mechanisms to obtain exact representations
from decimal import Decimal
f4 = Decimal('0.1') + Decimal('0.1') + Decimal('0.1') 
f4

Decimal('0.3')

If number types are mixed, Python will do the conversion. 

In [26]:
# Mix int and float; Python will convert int to float first 
a = 1 + 2.5 
print(a)
type(a)

3.5


float

In [27]:
# Convert between different types
a = 2
b = float(a)
print(b)
type(b)

2.0


float

In [28]:
c = int(b)
print(c)
type(c)

2


int

We can also use logic comparisons with numbers using `<`, `>`, `>=`, `<=`.

In [29]:
# Logic comparison
5<3

False

#### Built-in Mathematical Functions
Examples of *built-in mathematical functions* include: `pow`, `abs`, `round`, and others. These functions are built into the Python interpreter. We do not need to import any packages. Check the list of all built-in functions in Python:
https://docs.python.org/3.10/library/functions.html

In [30]:
# Power (exponentiation)
pow(2,4)

16

In [31]:
round(3.006)

3

In [32]:
# Absolute value
abs(-3.4) 

3.4

In [33]:
# Check documentation for help about built-in functions
help(pow) 

Help on built-in function pow in module builtins:

pow(base, exp, mod=None)
    Equivalent to base**exp with 2 arguments or base**exp % mod with 3 arguments

    Some types, such as ints, are able to use a more efficient algorithm when
    invoked using the three argument form.



#### Python Modules for Numerical Operations
The number of built-in mathematical functions is limited, and we can also import Python modules, such as `math` and `random` to perform mathematical operations. 
https://docs.python.org/3.10/library/math.html#module-math

In [34]:
import math

math.floor(3.006)

3

In [35]:
import random

# Return a random floating-point number in the range 0-1
r = random.random()
print(r)

0.18710060588576483


<a id="2.1.3-strings"></a>

### 2.1.3 Strings

> A **string** is an immutable sequence containing letters, words, and other characters.

Strings are used in Python to record text information, such as names. Strings in Python are *sequences*, which means that Python keeps track of every element in the string, and we can use indexing to get particular elements in the sequence.

#### Creating a String
To create a string in Python we can use either single quotes or double quotes.

In [36]:
# Single word
'hello'

'hello'

In [37]:
# Entire phrase 
'This is also a string'

'This is also a string'

In [38]:
# We can also use double quotes
"String built with double quotes"

'String built with double quotes'

Note that the code below shows an error, because the single quote in `I'm` broke the continuation of the single quotes in the string. 

In [39]:
# Be careful with quotes!
' I'm using single quotes, but this will create an error'

SyntaxError: unterminated string literal (detected at line 2) (2053197537.py, line 2)

You can use combinations of double and single quotes to get the complete statement.

In [40]:
"Now I'm ready to use the single quotes inside a string!"

"Now I'm ready to use the single quotes inside a string!"

#### Printing a String

In Jupyter notebooks, writing a string in a cell will automatically output the string, however the correct way to display strings is by using a **print** function.

In [41]:
# We can simply declare a string
'Hello World'

'Hello World'

In [42]:
# Note that we can't output multiple strings this way; only the last string is displayed
'Hello World 1'
'Hello World 2'

'Hello World 2'

We can use a `print` statement to display a string, or multiple strings in a cell.

In [43]:
print('Hello World 1')
print('Hello World 2')
print('Use \n to print a new line') # \n prints a new line
print('\n')
print('See what I mean?')

Hello World 1
Hello World 2
Use 
 to print a new line


See what I mean?



We can also use the built-in function `len()` to check the length of a string. It counts all of the characters in the string, including spaces and punctuation marks.

In [44]:
len('Hello World')

11

#### String Indexing and Slicing
Since strings are sequences, Python can use indexes to call parts of the sequence.

**Indexing** starts at 0 in Python. 

In [45]:
# Assign a string to the variable 's'  
s = 'Hello World'

In [46]:
# Show 's'
s

'Hello World'

In [47]:
# Print the string
print(s) 

Hello World


In [48]:
# Check the type of 's'
type(s)

str

In [49]:
# Show first element
s[0]

'H'

Use the slicing operator `:` to perform **slicing** which returns the elements up to a designated index in the sequence.

In [50]:
# Grab everything past the first element all the way to the end of s 
s[1:]

'ello World'

In [51]:
# Note that there is no change to the original s
s

'Hello World'

In [52]:
# Grab everything UP TO the 3rd index
s[:3]

'Hel'

The above slicing includes indexes 0, 1, and 2, and it doesn't include the 3rd index. In Python, slicing is performed as **up to, but not including**.

In [53]:
# Return everything
s[:]

'Hello World'

We can also use negative indexing to go backwards.

In [54]:
# Last letter (one index behind 0 so it loops back around)
s[-1]

'd'

In [55]:
# Grab everything but the last letter
s[:-1]

'Hello Worl'

We can also use indexing and slicing notation to grab elements of a sequence by a specified step size (the default is 1). For instance, we can use two colons in a row `::` and then a number specifying the frequency to grab elements. 

In [56]:
# Grab everything, but go in steps size of 1
s[::1]

'Hello World'

In [57]:
# Grab everything, but go in step sizes of 2
s[::2]

'HloWrd'

In [58]:
# We can use step size of -1 to print a string backwards
s[::-1]

'dlroW olleH'

#### String Properties
Strings are *immutable* objects. It means that once a string is created, the elements within it **can not** be changed or replaced. 

In [59]:
s

'Hello World'

In [60]:
# Let's try to change the first letter to 'x'
s[0] = 'x'

TypeError: 'str' object does not support item assignment

Other properties of strings include concatenation, i.e., we can concatenate strings.

In [ ]:
s

In [61]:
# Concatenate strings
s + ' concatenate me!'

'Hello World concatenate me!'

In [62]:
# We can reassign s completely
s = s + ' concatenate me!'

In [63]:
# Note that now s points to the entire sequence
print(s)

Hello World concatenate me!


We can also use the multiplication symbol `*` to create a repetition of a string.

In [64]:
letter = 'z'
letter*10

'zzzzzzzzzz'

#### Built-in Methods for Strings

Objects in Python can also have **built-in methods**. Differently from built-in functions in Python which can typically be used with different object types, built-in methods are specific to particular object types. Also, while built-in functions are called directly by their name (e.g., `len()`, or `type()`), build-in methods are called with a period followed by the method name, as in: 

`object.method(parameters)`

In the above line, `parameters` are extra arguments we can pass into the method.

Here are some examples of built-in methods for strings.

In [65]:
s

'Hello World concatenate me!'

In [66]:
# Upper case the string
s.upper()

'HELLO WORLD CONCATENATE ME!'

In [67]:
# Lower case
s.lower()

'hello world concatenate me!'

In [68]:
# Split a string by blank spaces (this is the default)
s.split()

['Hello', 'World', 'concatenate', 'me!']

In [69]:
# Split by a specific element (doesn't include the element that was split on)
s.split('W')

['Hello ', 'orld concatenate me!']

In [70]:
# Check all built-in methods for the string s
dir(s)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'removeprefix',
 'removesuffix',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'stri

<a id="2.1.4-lists"></a>

### 2.1.4 Lists

> A **list** is a mutable ordered sequence of elements, written as a series of items within square brackets.

Lists are the most general example of a *sequence* in Python. Unlike strings, they are mutable, meaning that the elements inside a list can be changed.

Lists are constructed with square brackets `[]` and commas separating every element in the list.

In [71]:
# Assign a list to the variable named 'my_list'
my_list = [1, 2, 3]

Lists can hold different object types. For example, the following list contains strings, integers, and float numbers.

In [72]:
my_list = ['A string', 23, 100.232, 'o']
my_list

['A string', 23, 100.232, 'o']

We can think of lists as arrays of references (pointers) to objects with allocated memory. Each element in the list is a reference to an object (such as numbers, strings, lists, etc.), and not the object itself.

Just like with strings, the built-in function `len()` returns the number of items in the sequence of the list.

In [73]:
len(my_list)

4

#### List Indexing and Slicing
Indexing and slicing work just like in strings.

In [74]:
my_list = ['one', 'two', 'three', 4, 5]

# Grab element at index 0
my_list[0]

'one'

In [75]:
# Grab everything UP TO index 3
my_list[:3]

['one', 'two', 'three']

We can also use + to concatenate lists, just like we did for strings.

In [76]:
my_list + ['new item']

['one', 'two', 'three', 4, 5, 'new item']

Note that the above operation doesn't actually change the original list.

In [77]:
my_list

['one', 'two', 'three', 4, 5]

To make the change permanent, we need to reassign the list.

In [78]:
# Reassign
my_list = my_list + ['add new item permanently']

In [79]:
my_list

['one', 'two', 'three', 4, 5, 'add new item permanently']

We can also use the operator `*` for a duplication method similar to strings.

In [80]:
# Make the list double
my_list * 2

['one',
 'two',
 'three',
 4,
 5,
 'add new item permanently',
 'one',
 'two',
 'three',
 4,
 5,
 'add new item permanently']

In [81]:
# Again, doubling is not permanent
my_list

['one', 'two', 'three', 4, 5, 'add new item permanently']

Lists indexing will return an error if there is no element at that index. For example:

In [82]:
my_list[100]

IndexError: list index out of range

#### Built-in Methods for Lists

There are parallels between arrays in other programming languages and lists in Python. Lists in Python tend to be more flexible than arrays in other languages for two reasons: they have no fixed size (meaning we don't have to specify the size of a list when we create one), and they have no fixed type constraint (like we've seen above we can mix different types of objects in a list).

Explained next are several built-in methods for lists.

In [83]:
# Create a new list
list1 = [1, 2, 3]

Use the **append** method to permanently add an item to the end of a list:

In [84]:
# Append
list1.append('append me!')

In [85]:
# Show
list1

[1, 2, 3, 'append me!']

Use **pop** to extract ("pop off") an item from the list. By default, pop takes off the last index (i.e., with index -1), but we can also specify which index to pop off. 

In [86]:
# Pop off the 0 indexed item
list1.pop(0)

1

In [87]:
# Show
list1

[2, 3, 'append me!']

In [88]:
# Assign the popped element; remember that the default popped index is -1
popped_item = list1.pop()

In [89]:
popped_item

'append me!'

In [90]:
# Show remaining list
list1

[2, 3]

We can insert and remove elements from a list.

In [91]:
c = ['a', 'b', 'c']

# Insert at index 0
c.insert(0, 'a0')
print(c)

['a0', 'a', 'b', 'c']


In [92]:
# Remove
c.remove('b')
print(c)

['a0', 'a', 'c']


Note that *remove* removes a specific value (e.g., the value `'b'` in the list `c`), while *pop* removes an item at a specific index.

In [93]:
# We can also use the statement del to remove the item at index 0
del c[0]
print(c)

['a', 'c']


We can use the **sort** and **reverse** methods with lists.

In [94]:
new_list = ['a','e','x','b','c']

In [95]:
# Show
new_list

['a', 'e', 'x', 'b', 'c']

In [96]:
# Use reverse to reverse order (this is permanent!)
new_list.reverse()

In [97]:
new_list

['c', 'b', 'x', 'e', 'a']

In [98]:
# Use sort to sort the list (in alphabetical order)
new_list.sort()

In [99]:
new_list

['a', 'b', 'c', 'e', 'x']

In [100]:
# For list of numbers, sorting is in ascending order
list_of_numbers = [2, 4, 3, 7, 1]
list_of_numbers.sort()
list_of_numbers

[1, 2, 3, 4, 7]

Two lists can be combined into a single list by the **zip** function.

In [101]:
a = [1, 2, 3, 4, 5]
b = [5, 4, 3, 2, 1]

print(zip(a,b))

To see the results of the **zip** function, in the next cell we convert the returned zip object into a list. Note that the **zip** function returns a list of tuples.  Each tuple represents a pair of items that the function zipped together. The order in the two lists was maintained.

In [102]:
print(list(zip(a,b)))

[(1, 5), (2, 4), (3, 3), (4, 2), (5, 1)]


Often the **zip** command is used inside of a for-loop. The following code shows how a for-loop can assign a variable to each collection that the program is iterating. 

In [103]:
a = [1, 2, 3, 4, 5]
b = [5, 4, 3, 2, 1]

for x,y in zip(a,b):
    print(f'{x} - {y}')

1 - 5
2 - 4
3 - 3
4 - 2
5 - 1


#### Nesting Lists
Python data structures support **nesting**, that is, we can have data structures within data structures. For example, a list inside a list is shown next.

In [104]:
# Let's make three lists
lst_1 = [1, 2, 3]
lst_2 = [4, 5, 6]
lst_3 = [7, 8, 9]

# Make a list of lists to form a matrix
matrix = [lst_1, lst_2, lst_3]

In [105]:
# Show
matrix

[[1, 2, 3], [4, 5, 6], [7, 8, 9]]

We can again use indexing to grab elements, but now there are two levels for the index: the items in the matrix object, and the items inside that list.

In [106]:
# Grab the first item in the matrix object
matrix[0]

[1, 2, 3]

In [107]:
# Grab the first item of the first item in the matrix object
matrix[0][0]

1

#### List Comprehensions
Python has an advanced feature called **list comprehensions**. This allows for quick construction of lists.

The basic syntax of list comprehensions is:

`newlist = [expression for item in list]`

It returns a new list by applying the expression to each item in the list, or any other iterable object. 

In [108]:
# Apply x+10 to each element in the list
list2 = [1, 2, 3]
new_list = [x+10 for x in list2] 
new_list

[11, 12, 13]

We can also add logic conditions to the list comprehensions as in the next example.

In [109]:
# Even numbers in the range 0-10 
even_numbers = [x for x in range(10) if x%2==0] 
even_numbers

[0, 2, 4, 6, 8]

In [110]:
# Squared elements in the even numbers list
even_squares = [x*x for x in even_numbers]
even_squares

[0, 4, 16, 36, 64]

<a id="2.1.5-dictionaries"></a>

### 2.1.5 Dictionaries

> A **dictionary** is an unordered and mutable Python container that stores mappings of unique keys to values.

We've been learning about *sequences* in Python so far, and now we're going to learn about *mappings* in Python. If you're familiar with other programming languages, you can think of dictionaries as hash tables. 

Mappings are a collection of objects that are stored by a unique *key*, unlike a sequence that stores objects by their relative position (index). This is an important distinction, since mappings won't retain the order of the elements because each object is defined by a key.

Therefore, a Python dictionary consists of a collection of **keys** and associated **values**. A colon `:` separates each key from its value, and the keys and values are provided inside curly braces `{}`. The keys must be unique, and can appear only once in a dictionary. Also, the keys must be immutable objects, such as strings, integers, floats, or tuples. The associated values can be almost any Python object, and there are no restrictions.

Dictionaries are mutable, therefore the elements can be changed, added, and removed. 

#### Constructing a Dictionary

In [111]:
# Make a dictionary with {} and : to signify a key and a value
my_dict = {'key1':'value1','key2':'value2'}

In [112]:
my_dict

{'key1': 'value1', 'key2': 'value2'}

In [113]:
# Call values by their key
my_dict['key2']

'value2'

Dictionaries are very flexible and they can hold various data types.

In [114]:
# The values can be any object type
my_dict = {1:101, 2:102, 3:103}

# Call an item
my_dict[2]

102

In [115]:
my_dict = {'key1':123, 'key2':[12,23,33], 'key3':['item0','item1','item2']}

# Let's call items from the dictionary
my_dict['key3']

['item0', 'item1', 'item2']

In [116]:
# We can call an index on that value
my_dict['key3'][0]

'item0'

In [117]:
# Can even call methods on that value
my_dict['key3'][0].upper()

'ITEM0'

We can change the values in a dictionary, as in the following cell.

In [118]:
# Subtract 123 from the value
print(my_dict['key1'])
my_dict['key1'] = my_dict['key1'] - 100

123


In [119]:
#Check
my_dict['key1']

23

The keys in Python dictionaries are immutable, and we cannot change them. However, we can remove a key using `pop()` or `del` as in lists, and afterward add a new key with its associated value.

In [120]:
# Delete the key-value pair for 'key3'
del my_dict['key3']

my_dict

{'key1': 23, 'key2': [12, 23, 33]}

In [121]:
# Remove the key-value pair for 'key1'
my_dict.pop('key1')
my_dict

{'key2': [12, 23, 33]}

We can also create new keys and values by assignment. For instance, if we start with an empty dictionary, we can continually add key-value pairs to it.

In [122]:
# Create a new dictionary
d = {}

In [123]:
# Create a new key through assignment
d['animal'] = 'Dog'

In [124]:
# Can do this with any object
d['answer'] = 42

In [125]:
# Show
d

{'animal': 'Dog', 'answer': 42}

#### Nesting with Dictionaries

Python has flexibility of nesting objects and calling methods on them. Let's see a dictionary nested inside a dictionary.

In [126]:
# Dictionary nested inside a dictionary nested inside a dictionary
d = {'key1':{'nestkey':{'subnestkey':32}}}

In [127]:
# Keep calling the keys
d['key1']['nestkey']['subnestkey']

32

#### Dictionary Built-In Methods

There are several built-in methods we can call on a dictionary. 

In [128]:
# Create a dictionary
d = {'key1':1,'key2':2,'key3':3}

In [129]:
# Method to return a list of all keys 
d.keys()

dict_keys(['key1', 'key2', 'key3'])

In [130]:
# Method to return all values
d.values()

dict_values([1, 2, 3])

In [131]:
# Method to return all keys and values pairs
d.items()

dict_items([('key1', 1), ('key2', 2), ('key3', 3)])

<a id="2.1.6-tuples"></a>

### 2.1.6 Tuples

> A **tuple** is a collection of objects which is ordered and immutable, and it is commonly written as a series of items in parentheses.

In Python, tuples are very similar to lists, with the main difference being that tuples are *immutable* sequences, unlike lists that are *mutable* sequences. Tuples are created similarly to lists, but with parentheses `()` instead of squared brackets `[]`.

The basic characteristics of tuples include:

- They are ordered collections of objects: like lists and strings, tuples are positionally ordered collections of objects (i.e., they are sequences) that maintain a left-to-right order among their elements.
- Are accessed by offset: like strings and lists, items in a tuple are accessed by positional offset (not by key); therefore, they support indexing and slicing.
- Tuples are immutable sequences: like strings and lists, tuples are sequences. However, unlike lists that are *mutable* sequences, tuples are *immutable* sequences (meaning they can not be changed in place). 
- Are fixed-length, heterogeneous, and arbitrarily nestable: because tuples are immutable, their size cannot be changed (without making a new copy). Tuples can hold any type of object, including other compound objects (e.g., lists, dictionaries, other tuples), and hence, they support arbitrary nesting.
- Tuples are arrays of object references: like lists, tuples are best thought of as arrays of references (pointers) to other objects with allocated memory.


#### Constructing Tuples

Tuples are constructed by using parentheses `()` with the items separated by commas. 

In [132]:
# Creating a tuple
t = (1, 2, 3)
t

(1, 2, 3)

In [133]:
# Check the length of the tuple using len(), just like a list
len(t)

3

In [134]:
# We can also mix object types: e.g., strings, integer numbers, floating-point numbers
t = ('one', 2, 490.2)

# Show
t

('one', 2, 490.2)

In [135]:
# Tuples, lists, or dictionaries can be nested into other tuples
w = ('one', 'two', (4, 5), 6, ['r', 100])
w

('one', 'two', (4, 5), 6, ['r', 100])

In [136]:
# An empty tuple
u = () 
u

()

In [137]:
# A 1-item tuple
v = ('thing', )    
v

('thing',)

Note that for a single-item tuple we need to place a comma after the item, that is, we use `(item,)` and not `(item)`, since parentheses can also be used to enclose expressions like `(1 + 2) * 3` = `9`.

In [138]:
# Note that the output of this cell is not a tuple
# Since the displayed output of the cell is not in parentheses, it is an integer number, not a tuple
a = (3)
a

3

We can also use the built-in function `type()` to check the type of the variable `a`.

In [139]:
# The type of the variable a is integer number
type(a)

int

In [140]:
# This is a tuple
b = (3,)
b

(3,)

In [141]:
# The type of the variable b is tuple
type(b)

tuple

In [142]:
# Not a tuple
(1 + 4) * 3

15

In [143]:
# This is a tuple: note that (1+4,) is the same as (5,), and when multiplied by 3, the tuple is repeated 3 times
(1 + 4,) * 3

(5, 5, 5)

The parentheses `()` can be omitted in the syntax, and tuples in Python can be created just by listing items separated with commas. Although the parentheses are mostly optional with tuples, there are a few cases when using parentheses is required, e.g., within a function call, or when nested in a larger expression. For beginners, it is recommended to always use parentheses, in order to avoid the above exceptions, and because they improve the code readability.

In [144]:
t = 'one', 2, 490.2
t

('one', 2, 490.2)

In [145]:
# A tuple with one item can be created just by adding a comma after the item without using parentheses
'hello',

('hello',)

#### Tuple Indexing and Slicing

Since tuples are positionally ordered collections of objects like strings and lists, indexing and slicing work for tuples.

In [146]:
t

('one', 2, 490.2)

In [147]:
# Use indexing just like in lists and strings
t[0]

'one'

In [148]:
t[1]

2

In [149]:
t[-1]

490.2

In [150]:
# Slicing
t[0:2]

('one', 2)

Other sequencing operations, such as concatenation and repetition, are also supported for tuples, in a similar way as for lists and strings.

In [151]:
# Concatenation
(1, 'book') + ('notes', 4) 

(1, 'book', 'notes', 4)

In [152]:
# Repetition
(1, 'thing') * 4 

(1, 'thing', 1, 'thing', 1, 'thing', 1, 'thing')

Because tuples are sequences, we can also use `for` loop iterations and list comprehensions to print the elements of tuples. 

In [153]:
# Consider the following tuple
x = ('b', 'u', 'i', 'l', 'd', 'i', 'n', 'g')
x

('b', 'u', 'i', 'l', 'd', 'i', 'n', 'g')

In [154]:
# We can use a `for` loop iteration to print each of the items of the tuple on a separate line
for i in x:
    print(i)

b
u
i
l
d
i
n
g


In [155]:
# A list comprehension can also be used to print each of the items of the tuple x on a separate line
l = [print(i) for i in x]

b
u
i
l
d
i
n
g


#### Built-in Methods for Tuples

There are built-in methods for tuples in Python, but not as many as for lists. Tuples do not have methods such as `append()`, `remove()`, `extend()`, `insert()`, and `pop()` due to their immutable nature.

In [156]:
# Show
t

('one', 2, 490.2)

In [157]:
# Use .index to enter an item and return the index
t.index('one')

0

In [158]:
# Use .count to count the number of times a value appears
t.count('one')

1

In [159]:
# Count the number of times 2 appears in the tuple
u = (1, 2, 3, 2, 1, 2)
u.count(2)

3

#### Tuple Immutability

To emphasize one more time that tuples are immutable, check the following examples.

In [160]:
# If we try to change the first element, we will get an error message
t[0] = 'four'

TypeError: 'tuple' object does not support item assignment

Because of their immutability, tuples can't grow. Once a tuple is created, we can not add to it.

In [161]:
# We will get an error message
t.append('nope')

AttributeError: 'tuple' object has no attribute 'append'

We can, however, make a new tuple based on a current tuple.

In [162]:
t = (t[0], 7, t[2])
t

('one', 7, 490.2)

#### Conversion to Lists

Conversion to lists and back to tuples is straightforward.

In [163]:
# Tuple to list
l = list(t)
l

['one', 7, 490.2]

In [164]:
type(l)

list

In [165]:
# List to tuple
l2 = ['aa', 'bb', 5, 'cc']
t2 = tuple(l2)
t2

('aa', 'bb', 5, 'cc')

#### Tuple Unpacking

Tuple **unpacking** means pairing objects on the right side of the assignment operator `=` with targets on the left side by position, and assigning them from left to right.

In [166]:
# Unpacking the tuple into the individual items
y = ('GOOG', 120, 490.2)
order, shares, price = y
print(order)
print(shares)
print(price)

GOOG
120
490.2


In [167]:
print('Cost:', shares * price)

Cost: 58824.0


In [168]:
# Unpacking the tuple: two names are entered for a tuple with 3 items, resulting in an error
order, shares = y

ValueError: too many values to unpack (expected 2)

#### Named Tuples

Named tuples are an extended type of tuples that allow items to be accessed by both position and attribute name, similar to dictionaries. They are created by using the `namedtuple` function from the `collections` module.

In [169]:
# Import and create a named tuple
from collections import namedtuple 

Rec = namedtuple('Record', ['name', 'age', 'jobs']) 

# Assign a named-tuple record
bob = Rec(name='Bob', age=40.5, jobs=['dev', 'mgr']) 
alice = Rec(name='Alice', age=36, jobs=['mgr']) 

In [170]:
# Access by position
bob[0]

'Bob'

In [171]:
bob[1]

40.5

In [172]:
# Access by attribute 
bob.name, bob.jobs 

('Bob', ['dev', 'mgr'])

A named tuple can be converted to a dictionary, which allows key-based access to the items.

In [173]:
D = bob._asdict() 
D

{'name': 'Bob', 'age': 40.5, 'jobs': ['dev', 'mgr']}

In [174]:
# Access by key
D['name']

'Bob'

#### When to Use Tuples

Although tuples are very similar to lists, tuples are not used as often as lists in programming. However, tuples are used when immutability is necessary; for instance, if in your program you are using an object and need to make sure it does not get changed, then a tuple provides convenient integrity. 

<a id="2.1.7-sets"></a>

### 2.1.7 Sets

>  A **set** is a collection of *unique* objects which is unordered and mutable, and are constructed by using the `set()` function.

Sets support operations corresponding to mathematical set theory, such as intersection, union, etc. By definition, an item appears only once in a set, no matter how many times it is added. 

Because sets are collections of objects, they share some behavior with lists and dictionaries. For example, sets are iterable, can grow and shrink on demand, and may contain a variety of object types. 

However, since sets are unordered and do not map keys to values, they are neither a sequence nor mapping type.

Sets have a variety of applications, especially in numeric and database-focused work.

To create a set object, pass in a sequence or another iterable object to the built-in set function.

In [175]:
# Create a set from string
x = set('abcde')
x

{'a', 'b', 'c', 'd', 'e'}

In [176]:
# Create a set from list
x = set([2,3,4,5])
x

{2, 3, 4, 5}

In [177]:
# Create a set from tuple
x = set((5,6,7,8))
x

{5, 6, 7, 8}

The sets are displayed with curly brackets. This is similar to a dictionary, but sets do not have keys and values (or, they can be considered dictionaries with only keys and without any values).

#### Set Expressions

In [178]:
x = set('abcde')
y = set('bdxyz')

In [179]:
# Union
x | y

{'a', 'b', 'c', 'd', 'e', 'x', 'y', 'z'}

In [180]:
# Intersection
x & y

{'b', 'd'}

In [181]:
# Difference
x - y

{'a', 'c', 'e'}

In [182]:
# Symmetric difference (XOR) - elements in either x or y, but not both in x and y
x ^ y

{'a', 'c', 'e', 'x', 'y', 'z'}

In [183]:
# Superset, subset
x > y, x < y

(False, False)

In [184]:
# Membership of a set
'e' in x

True

Sets can also be created by adding elements to an existing set object.

In [185]:
# Create a set
z = set()

In [186]:
# Add to set with the add() method
z.add(1)
z

{1}

In [187]:
# Add a different element
z.add(2)
z

{1, 2}

In [188]:
# Try to add the same element
z.add(1)
z

{1, 2}

We cannot add another 1, because a set has only unique elements.

For instance, we can cast a list with multiple repeat elements into a set to get the unique elements of the list.

In [189]:
# Create a list with repeats
list1 = [1, 1, 2, 2, 3, 4, 5, 6, 1, 1]
# Cast as set to get unique values
set(list1)

{1, 2, 3, 4, 5, 6}

#### Built-in Methods for Sets

Similar to the set expressions shown above, there are built-in set methods for union, intersection, and other related operations.

In [190]:
x = set('abcde')
y = set('bdxyz')
# Same as x & y
z = x.intersection(y)
z

{'b', 'd'}

In [191]:
# Delete one item
z.remove('b') 
z

{'d'}

Also, we can use for-loops with the elements of sets.

In [192]:
for item in set('abc'):
    print(item * 3)

aaa
ccc
bbb


<a id="2.1.8-other-data-types"></a>

### 2.1.8 Other Data Types

#### Booleans

Python also has a Boolean data type with predefined built-in names `True` and `False`, that are basically just the integers 1 and 0.

In [193]:
# Assign the Boolean True object to 'a'
a = True
# Show
a

True

The data type for `True` and `False` is `bool`.

In [194]:
type(a)

bool

We can also use comparison operators to create Booleans. 

In [195]:
# Output is boolean
1 > 2

False

In [196]:
# Is True the same as 1
True == 1

True

In Python each object is either True or False, as follows:

- Numbers are false if zero, and true otherwise.
- Other objects are false if empty, and true otherwise.

In [197]:
bool(2)

True

In [198]:
bool(0)

False

In [199]:
bool('book')

True

In [200]:
bool('')

False

In [201]:
bool([1, 2])

True

#### The None Object

We can use **None** as a placeholder for an object that we don't want to reassign yet.

In [202]:
# None placeholder
b = None

In [203]:
# Show
print(b)

None


For instance, to initialize a list whose size is not known yet, we can use `None` to preset the initial size and allow for future index assignment.


In [204]:
# Initialize a list of size 5 with None
my_list = [None] * 5

# Show
my_list

[None, None, None, None, None]

In [205]:
# Assign strings to the first two elements
my_list[0] = 'apple'
my_list[1] = 'banana'

# Show
my_list

['apple', 'banana', None, None, None]

Note also that in Python the fixed values or the raw data that are assigned to variables or constants are called **literals**. Examples of Python literals include: numeric literals (e.g., the integer number `5`, or float number `2.1`), string literals (e.g., any string like `'hi'` or `"hello"`), Boolean literals (`True` and `False`), special literal `None`, list literals (e.g., the list `[1, 2, 3]`), etc. Therefore, a literal is a sequence of characters or a single character that represents a fixed value in source code. Or, literals are the notation that we use to specify values in our code, which is converted by the Python interpreter into an object of the appropriate type.

<a id="2.1.9-string-formatting"></a>

### 2.1.9 String Formatting

String formatting allows injecting items into a string, rather than trying to chain items together using commas or string concatenation. As a quick comparison, consider:

    player = 'Thomas'
    points = 33

    # concatenation
    'Last night, '+player+' scored '+str(points)+' points.'  

    # string formatting
    f'Last night, {player} scored {points} points.'          
    
    # The output of both concatenation and string formatting is the same:
    'Last night Thomas scored 33 points.'

There are three ways to perform string formatting.

* The oldest method involves placeholders using the modulo `%` character.
* An improved technique uses the `.format()` string method.
* The newest method, introduced with Python 3.6, uses formatted string literals, called `f-strings`.

These three methods are described next.

#### Formatting with Placeholders
We can use `%s` to inject strings into print statements. The modulo `%` is referred to as a **string formatting operator**.

In [206]:
print("I'm going to inject %s here." %'something')

I'm going to inject something here.


We can pass multiple items by placing them inside a tuple after the `%` operator.

In [207]:
print("I'm going to inject %s text here, and %s text here." %('some','more'))

I'm going to inject some text here, and more text here.


We can also pass variable names.

In [208]:
x, y = 'some', 'more'
print("I'm going to inject %s text here, and %s text here."%(x,y))

I'm going to inject some text here, and more text here.


Note that there are two notations `%s` and `%r` that convert any Python object to a string using two separate methods: `str()` and `repr()`. Here, `%r` and `repr()` deliver the official or formal or raw string representation of the object, including quotation marks and any escape characters, wehreas `%s` and `str()` produce a readable of informal and user-friendly representation of an object. The `repr()` can be used to recreate the object, and is typically intended for debugging purposes.

In [209]:
print('He said his name was %s.' %'Fred')
print('He said his name was %r.' %'Fred') 
# Note that in the output 'Fred' is displayed in quotations

He said his name was Fred.
He said his name was 'Fred'.


As another example, `\t` inserts a tab into a string. Note that `%r` output the raw string representation and ignored the slash `\` in `\t`, therefore a tab was not inserted.

In [210]:
print('I once caught a fish %s.' %'this \tbig')
print('I once caught a fish %r.' %'this \tbig')

I once caught a fish this 	big.
I once caught a fish 'this \tbig'.


The `%s` operator converts whatever it sees into a string, including integers and floats. Similarly, the `%d` operator converts numbers to integers. Note the difference below.

In [211]:
print('I wrote %s programs today.' %3.75)
print('I wrote %d programs today.' %3.75)   

I wrote 3.75 programs today.
I wrote 3 programs today.


##### Padding and Precision of Floating Point Numbers
Floating point numbers use the format `%5.2f`. Here, `5` is the minimum number of characters the string should contain, including the decimal point. These characters may be padded with whitespace if the entire number does not have this many digits. Next to this, `.2f` stands for how many numbers to show past the decimal point. 

In [212]:
print('Floating point numbers: %5.2f' %(13.148))

Floating point numbers: 13.15


In [213]:
print('Floating point numbers: %1.0f' %(13.148))

Floating point numbers: 13


In [214]:
print('Floating point numbers: %1.5f' %(13.148))

Floating point numbers: 13.14800


In [215]:
# Note that 5 empty spaces will be added in front of the number, to make it a total of 10 characters
print('Floating point numbers: %10.2f' %(13.148))

Floating point numbers:      13.15


In [216]:
print('Floating point numbers: %25.2f' %(13.148))

Floating point numbers:                     13.15


It is possible to use more than one operator (e.g., `%s`, `%f`, and `%r`) in the same print statement.

In [217]:
print('First: %s, Second: %4.2f, Third: %r' %('hi!',3.1415,'bye!'))

First: hi!, Second: 3.14, Third: 'bye!'


#### Formatting with the `.format()` Method
An improved way to format objects into strings for print statements is with the string `.format()` method. The syntax is:

    'String here {} then also {}'.format('something1','something2')
    
For example:

In [218]:
print('This is a string with an {}'.format('insert'))

This is a string with an insert


The `.format()` method has several advantages over the `%s` placeholder method:

1. Inserted objects can be called by index position.

In [219]:
print('The {2} {1} {0}'.format('fox','brown','quick'))

The quick brown fox


2. Inserted objects can be assigned keywords.

In [220]:
print('First Object: {a}, Second Object: {b}, Third Object: {c}'.format(a=1, b='Two', c=12.3))

First Object: 1, Second Object: Two, Third Object: 12.3


3. Inserted objects can be reused, avoiding duplication.

In [221]:
print('A %s saved is a %s earned.' %('penny','penny'))
# vs.
print('A {p} saved is a {p} earned.'.format(p='penny'))

A penny saved is a penny earned.
A penny saved is a penny earned.


Within the curly braces we can assign field lengths, left/right alignments, rounding parameters, and more.

In [222]:
# The field 0 has a length of 8 characters, and the next field 1 has a length of 10 characters
print('{0:8} | {1:10}'.format('Fruit', 'Quantity'))
print('{0:8} | {1:10}'.format('Apples', 3.))
print('{0:8} | {1:10}'.format('Oranges', 10))

Fruit    | Quantity  
Apples   |        3.0
Oranges  |         10


By default, `.format()` aligns text to the left, numbers to the right. We can pass an optional `<`,`^`, or `>` to set a left, center, or right alignment, respectively.

In [223]:
print('{0:<8} | {1:^10} | {2:>8}'.format('Left','Center','Right'))
print('{0:<8} | {1:^10} | {2:>8}'.format(11,22,33))

Left     |   Center   |    Right
11       |     22     |       33


We can precede the alignment operator with a padding character.

In [224]:
print('{0:=<8} | {1:-^10} | {2:.>8}'.format('Left','Center','Right'))
print('{0:=<8} | {1:-^10} | {2:.>8}'.format(11,22,33))

Left==== | --Center-- | ...Right
11====== | ----22---- | ......33


Field widths and float precision are handled in a way similar to placeholders. The following two print statements are equivalent.

In [225]:
print('This is my ten-character, two-decimal number:%10.2f' %13.579)
print('This is my ten-character, two-decimal number:{0:10.2f}'.format(13.579))

This is my ten-character, two-decimal number:     13.58
This is my ten-character, two-decimal number:     13.58


Note that there are 5 spaces following `number:` in the output, and 5 characters are taken up by 13.58, for a total of 10 characters. In `{0:10.2f}`, the number `0` is a positional argument index which indicates to use the first number in `.format()`; if there were more than one number in `.format()`, we can use different numbers to refer to each one. Similarly, `10` indicates the minumum number of characters, `2` is the precision specifying to format the output with two decimal places, and `f` indicates that the output is a float number.

#### Formatting with String Literals (f-strings)

Introduced in Python 3.6, `f-strings` offer several benefits over the older `.format()` string method described above. E.g., we can bring outside variables immediately into the string rather than pass them as arguments through `.format(var)`.

In [226]:
name = 'Fred'

print(f"He said his name is {name}.")

He said his name is Fred.


Pass `!r` to call the raw string representation with the `repr()` method.

In [227]:
print(f"He said his name is {name!r}")

He said his name is 'Fred'


Float formatting follows the syntax `{value:{width}.{precision}}`.

Note the difference: with the `.format()` method the expression `{value:10.4f}`, with f-strings becomes `{value:{10}.{6}}`.

In [228]:
print("My 10 character, four decimal number is:{0:10.4f}".format(23.45678))
print(f"My 10 character, four decimal number is:{23.45678:{10}.{6}}")
print(f"My 10 character, two decimal number is:{23.45678:{10}.{4}}")

My 10 character, four decimal number is:   23.4568
My 10 character, four decimal number is:   23.4568
My 10 character, two decimal number is:     23.46


Note that with f-strings, *precision* refers to the total number of digits, not just those following the decimal point. This fits more closely with scientific notation and statistical analysis. 

Unfortunately, f-strings do not pad to the right of the decimal point, even if precision allows it. In the following cell, zeros are not added to `23.47` although we set the precision to 6 (sigh).

In [229]:
print(f"My 10 character, two decimal number is:{23.47:{10}.{6}}")

My 10 character, two decimal number is:     23.47


If this becomes important, we can always use `.format()` method syntax inside an `f-string`.

In [230]:
print(f"My 10 character, four decimal number is:{23.47:10.4f}")

My 10 character, four decimal number is:   23.4700


<a id="2.2-statements"></a>

## 2.2 Statements

Python code can be decomposed into **packages**, **modules**, **statements**, and **expressions**, as follows:

1. Packages are composed of modules.
2. Modules contain statements.
3. Statements contain expressions.
4. Expressions create and process objects.

> **Expressions** are part of statements that return a value, such as variables, operators, or function calls. 

For example, expressions in Python include `2+5`, or `x+3`, or `x * y` for given `x` and `y`, or `func1(3)` for a given function `func1`, since they all return a value. Expressions perform operations upon objects, and produce a value that can be used in other operations, can be assigned to a variable, printed out, etc. An expression can be part of larger expressions of statements. For instance, in the statement `a = 2 + 5`, the part `2 + 5` is an expression.

> **Statements** are sections of code that perform an action. The main groups of Python statements are: assignment statements, print statements, conditional statements (`if`, `break`, `continue`, `try`), and looping statements (`for`, `while`). 

Therefore, `a = 2 + 5` is a statement, because it doesn't return a value, but it performs an action by assigning the value of `2 + 5` to the variable `a`. Beside assignment statements (such as `a = 2 + 5`), other forms of statements include conditional statements (e.g., `if x > 5:`), looping statements (e.g., `for i in range(10):`), print statements (e.g., `print('hi')`), as well as there are several other statements, such as `import`, `return`, `pass`, etc. Note again that none of these statements returns a value in the way expressions do. 

> **Modules** are Python files that contain Python statements, and are also called scripts.

Thus, a module is a single Python script that is composed of a series of related statements grouped into one file. The statements in a module can assign values to variables, create functions or classes, or perform other actions.  

> **Packages** are Python programs that collect related modules together within a single directory hierarchy.

In other words, a package is a directory that contains a collection of module files. Packages can also contain sub-packages, forming a hierarchy of packages. A Python program can be organized into a single package, or more complex programs can use multiple packages to achieve its functionality.

In this lecture, we will study several types of Python statements. In a subsequent lecture, we will provide explanations about modules and packages in Python. 

<a id="2.2.1-if,-else,-elif-statements"></a>

### 2.2.1 if, else, elif Statements

In Python, the `if` statement allows to instruct the program to perform alternative actions, based on one or several tests. This provides a means for introducing logic in our codes, and it can be interpreted as "if this case happens, then perform this action".

The `if` statement takes the form of an `if test`, which can be followed by one or more optional `elif (else if) tests`, and a final optional `else test`. Each of the tests has an associated block of nested statements, indented under a header line.

The `if` statement is a *compound statement,* since it may contain other statements (e.g., `elif` or `else`) in its syntax. 

> **Compound statements** in Python contain other statements, and they affect or control the execution of those other statements in some way. Compound statements typically span multiple lines.

Also, the `if` statement is referred to as a **conditional statement**, since it involves actions that are performed only when the conditions in the `if test` are satisfied.

#### Basic if Test

In its simplest form, the `if` statement has the following syntax:
    
    if test1:
        code to execute when test1 is True

The `if` statement is used to perform a test and control whether or not the indented block of code is executed.

- The first line is a header line. It opens with the `if` keyword, followed by a test condition, and ends with a colon `:` (omitting the colon at the end of `if` statement is one of the most common mistakes by beginner programmers in Python). The output of the expression in the `if` test is a *Boolean* variable (i.e., *True* or *False*).   
- The block of code is indented under the header and contains one or more statements (such as the `print` statements in the next cell) that are executed if the test is *True*.

In [1]:
x = 105

if x > 100:
    print(x, 'is high')

105 is high


In [2]:
x = 105

# The print statement is not executed since x<50 is False
if x < 50:
    print(x, 'is high')

In [3]:
y = 20

if y < 50:
    print (y, 'is low')

20 is low


In [4]:
if True:
    print('It is true!')

It is true!


In [5]:
# The print statement is not executed
if False:
    print('It is true!')

Therefore, the code indented under an `if` line will be executed only if the first line returns a Boolean `True` value. As we mentioned earlier, any nonzero number or nonempty array returns a Boolean `True`, and 0 or an empty array returns a `False`.

In [6]:
if 1:
    print('It is true!')
    
if 5:
    print('It is also true!')

# The print statement is not executed
if 0:
    print('It is not true!')

It is true!
It is also true!


#### The if - else Tests

The `else` test allows to add additional logic to the `if` test. 

Check the example in the following cell. Since we assigned a Boolean `False` to the variable `x`, the line `if x:` returns *False*, and as a result the statement indented under `if` will not be executed. In the case when the `if` test is *False*, the code after `else` is executed.

In [7]:
x = False

if x:
    print('This is printed when x is True!')
else:
    print('This is printed when x is False')

This is printed when x is False


Here are more examples of using `else` to execute a block of code when an `if` test is *not* true.

In [8]:
num = 43

if num > 100:
    print(num, 'is high')
else:
    print(num, 'is low')

43 is low


In [9]:
num = 134

if num > 100:
    print(num, 'is high')
else:
    print(num, 'is low')

134 is high


In the next example we use the `input()` function to enter text using the keyboard (press the `Enter` key to confirm it).

In [10]:
person = input("Enter your name: ")
# E.g., enter other name than Joe

if person == 'Joe':
    print('Welcome Joe!')
else:
    print("Welcome, Joe will be with you shortly?")

Enter your name:  Jim


Welcome, Joe will be with you shortly?


 Note that:

- `else` is always attached to `if`, and it cannot be used as a standalone test.
- `else` allows to specify an alternative action to execute when the `if` test is `False`.

#### if - elif - else Tests

We can use `elif` to specify additional tests, when we want to provide several alternative cases, each with its own test. The statement `elif` is short for "else if" and it is always associated with an `if` statement. If there is an `else` test in the code, `elif` must come before `else`.

The general syntax is:

    if test1:
        code to execute -> perform action 1
    elif test2:
        code to execute -> perform action 2
    else: 
        code to execute -> perform action 3
        
The above compound statement can be interpreted as: if the case in test 1 happens, perform action 1. Else, if the case in test 2 happens, perform action 2. Or else, if none of the above cases happen, perform action 3.

That is, Python executes the statements nested under the `elif` test if the statements before the test are not *True*, and the statements under the `else` test are executed only when none of the `elif` tests is *True*.

Both the `elif` and `else` parts are optional and they may be omitted. As well as, there may be more than one `elif` statement nested in the `if` test.

The words `if`, `elif`, and `else` line up vertically, and need to have the same indentation. The block of code under each test needs to be aligned if it consists of multiple lines of code, however the code under `if` does not have to have the same indentation as the code under `elif` or `else` (although it is recommended to use consistent indentation).

In [11]:
z = 68
    
if z > 100:
    print(z, 'is high')
elif z > 50:
    print(z, 'is medium')
else:
    print(z, 'is low')

68 is medium


In [12]:
z = 30
    
if z > 100:
    print(z, 'is high')
elif z > 50:
    print(z, 'is medium')
else:
    print(z, 'is low')

30 is low


In [13]:
location = 'Bank'

if location == 'Auto Shop':
    print('Welcome to the Auto Shop!')
elif location == 'Bank':
    print('Welcome to the bank!')
else:
    print('Where are you?')

Welcome to the bank!


#### Boolean Operators to Make Complex Statements

We can create complex conditional statements with Boolean operators like **and** and **or**, or use comparators like `<`, `>`, or other comparators.

In [14]:
age = 40

if age > 65 or age < 16:
    print(age, 'is outside the labor force')
else:
    print(age, 'is in the labor force')

40 is in the labor force


We saw in the examples above that we can use double equal signs `==` to check if two objects are the same. Similarly, we can use an exclamation point and equal sign `!=` to check if two objects are not the same. 

In [15]:
person = 'Jim'

if person != 'Joe':
    print("Welcome, what's your name?")
else:
    print('Welcome Joe!')

Welcome, what's your name?


#### The if - else Ternary Expression

Python also has an **if - else ternary expression** with the following syntax:

    action_a if condition else action_b
    
In the above expression, first the `condition` is evaluated, and afterward either `action_a` or `action_b` is executed based on the Boolean value of the condition.

Let's reconsider the `if-else` statement example that we saw earlier.

In [16]:
num = 43

if num > 100:
    print(num, 'is high')
else:
    print(num, 'is low')

43 is low


The corresponding `if-else` ternary expression is as follows. 

In [17]:
print(num, 'is high') if num > 100 else print(num, 'is low')

43 is low


The ternary expression allows to reduce the above 4 lines of code into 1 line. Based on the value of the condition `num > 100`, if the condition is True then `print(num, 'is high')` is executed, and if the condition is False then `print(num, 'is low')` is executed.

#### Handling Case Switch

If you used languages like C, Pascal, or MATLAB, and if you are interested to know if there is a *switch* or *case* statement in Python that selects an action based on a variable’s value, there isn't. Instead, in Python we can code multiway branching as a series of if-elif tests.

An example is shown below. Note again that we can use as many `elif` statements as we want, but there can be only one `else` statement.

In [18]:
choice = 'ham'

if choice == 'spam': 
    print(2.25)
elif choice == 'ham':
    print(1.75)
elif choice == 'eggs':
    print(0.75)
elif choice == 'bacon':
    print(1.10)
else:
    print('Bad choice')

1.75


Although, it may be more convenient to create a dictionary to handle case switching instead of `if-elif-else` especially when there are many cases involved. 

In [19]:
branch = {'spam': 2.25, 'ham': 1.75, 'eggs': 0.75, 'bacon': 1.10}

choice = 'eggs'

if choice in branch:
    print(branch[choice])
else:
    print('Bad choice')

0.75


#### Indentation Rules

Python uses indentation of statements under a header to group the statements in a nested block. In the figure below, there are 3 blocks of code, each having a header line. Note that Block 1 is nested under Block 0, and it is indented further to the right of Block 0. Then, Block 2 is nested under Block 1, and it is intended even further to the right of Block 1.

<img src="images/pic1b.jpg" width="270">
<em>Figure source: Reference [1].</em>

The indentation in Python is used to detect block boundaries. All statements indented the same distance to the right belong to the same block of code. The block ends either when a less-indented line or the end of the file is encountered. 

Indentation may consist of any number of spaces, but it must be the same for all the statements in a single block. Four spaces or one tab per indentation level are commonly used, but there is no absolute standard for the number of spaces in indentation. However, it is not recommended to mix spaces and tabs for indentation within a block, because such indentation may look different in other editors and the codes can be more difficult to edit. 

Look at the example in the next cell. It contains three blocks: the first block (Block 0, `if x:`) is not indented at all, the second (Block 1, `y = 2`) is indented four spaces under Block 0, and the third (Block 2, `print ('Block 2')` is indented eight spaces.

In [20]:
x = 1
if x:
    y = 2
    if y:
        print('Block 2')
    print('Block 1')
print('Block 0')

Block 2
Block 1
Block 0


Several common mistakes with code indentation are shown below, which result in errors.

In [22]:
x = 1
  if x: # Error: first line indented, this line belongs to Block 0 and it shouldn't be indented
    y = 2
      if y:  # Error: unexpected indentation, this line should have the same indentation as 'y = 2'
        print('Block 2')
   print('Block 1') # Error: inconsistent indentation, this line is indented 3 spaces, and 'y = 2' is indented 4 spaces
print('Block 0')

IndentationError: unindent does not match any outer indentation level (<tokenize>, line 6)

To indent several lines of code for one tab, select the lines and then press either the `Tab` key or press the keys `Ctrl` + `]`. To unindent several lines of codes for one tab, press the keys `Ctrl` + `[`.

#### Statement Delimiters: Lines and Continuations

Python expects `if` statements to be written on a single line.

The code below produces an error because the `if` statement spans two lines.

In [23]:
num = 80

if num > 20 and num > 50 and 
    num < 200 and num < 100:
    print('Medium number')

SyntaxError: invalid syntax (695293465.py, line 3)

When a statement is too long to fit on a single line, there are two ways to make it span multiple lines.

The first one is to enclose the statement either in a pair of parentheses `()`, square brackets `[]`, or curly braces `{}`. Continuation lines do not need to be indented at any level, but it is a good practice to align the lines vertically for readability.

Examples are shown below.

In [25]:
num = 80

if (num > 20 and num > 50 and 
    num < 200 and num < 100):
    print('Medium number')

Medium number


In [26]:
# Note that the indentation is not required for continuation lines enclosed in a pair of parentheses, brackets, or braces
num = 80
if {num > 20 and num > 50 and 
  num < 200 and num < 100}:
    print('Medium number')

Medium number


Also, statements can span multiple lines if they end in a backward slash `\`. Although, this is an older feature, and it is not generally recommended. One reason is because if there are empty spaces after the backward slash, it will result in an error.

In [27]:
num = 80
if num > 20 and num > 50 and \
    num < 200 and \
    num < 100:
    print('Medium number')

Medium number


The above line continuation rules apply to any other statements and expressions.

In [28]:
x = 1 + 2 + 3 \
+4
x

10

In [29]:
x = (1 + 2 + 3 
+4)
x

10

Note also that Python allows to write more than one noncompound statement (i.e., statements without nested statements) on the same line, separated by semicolons.

In [30]:
x = 5; print(x)

5


Python allows to write the body of a compound statement (like `if`) on the same line with the header, provided the body contains just simple (noncompound) statements (i.e., without `elif` or `else` tests).

In [31]:
if True: print('Something')

Something


<a id="2.2.2-for-loops"></a>

### 2.2.2 for Loops

A `for` loop acts as an iterator in Python. It goes through items that are in a *sequence* or any other iterable object. Objects that we've learned about that we can iterate over include strings, lists, and tuples. And even dictionaries allow to iterate over keys or values.

The general format of a `for` loop in Python is:

    for item in object:
        code to execute 

The variable name used for the `item` is completely up to the coder, so use your best judgment for choosing a name that makes sense and you will be able to understand when revisiting your code. This `item` can then be referenced inside your loop, for example if you wanted to use <code>if</code> statements to perform checks.

In [32]:
list1 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [33]:
for num in list1:
    print(num)

1
2
3
4
5
6
7
8
9
10


Add an <code>if</code> statement to check for even numbers. 

In [34]:
for num in list1:
    if num % 2 == 0:
        print(num)

2
4
6
8
10


We could have also included an `else` statement.

In [35]:
for num in list1:
    if num % 2 == 0:
        print(num)
    else:
        print('Odd number')

Odd number
2
Odd number
4
Odd number
6
Odd number
8
Odd number
10


Another common practice with `for` loops is to keep some sort of running tally during multiple loops. For example, let's create a `for` loop that sums up the elements in a list.

In [36]:
# Start sum at zero
list_sum = 0 

for num in list1:
    list_sum = list_sum + num

print(list_sum)

55


Or, we could have used the operator `+=` to perform the addition towards the sum. 

In [37]:
# Start sum at zero
list_sum = 0 

for num in list1:
    list_sum += num

print(list_sum)

55


We can also use `for` loops with strings and tuples, since they are sequences, so when we iterate through them we will be accessing each item in the sequence.

In [38]:
for letter in 'This is a string.':
    print(letter)

T
h
i
s
 
i
s
 
a
 
s
t
r
i
n
g
.


In [39]:
# loop through a dictionary
d = {'k1':1, 'k2':2, 'k3':3}

In [40]:
for item in d:
    print(item)

k1
k2
k3


Notice how the above produces only the keys. 

We can also use the dictionary built-in methods **.keys()**, **.values()**, and **.items()** with `for` loops. In Python each of these methods returns a *dictionary view object*. The view objects provide a view of the dictionary's key, values, and items (pairs of keys and values). The dictionary view objects support operations like membership tests and iterations over the keys, values, and items. The type of the view objects is `dict_items`. If we make changes to the dictionary, the view objects will keep track of the changes. 

In [41]:
# Create a dictionary view object
d.items()

dict_items([('k1', 1), ('k2', 2), ('k3', 3)])

Since the `.items()` method supports iteration, we can print both the keys and values.

In [42]:
# Dictionary unpacking
for k,v in d.items():
    print(k)
    print(v) 

k1
1
k2
2
k3
3


If we want to obtain a list of keys, values, or key-value tuples, we can *cast* the dictionary view objects as a list.

In [43]:
list(d.keys())

['k1', 'k2', 'k3']

In [44]:
# Compare to 
d.keys()

dict_keys(['k1', 'k2', 'k3'])

In [45]:
list(d.values())

[1, 2, 3]

In [46]:
list(d.items())

[('k1', 1), ('k2', 2), ('k3', 3)]

Another used function is **range** which allows to quickly *generate* a list of integers, and it is often used with `for` loops.

In [47]:
string = 'abcde'

n = len(string)
for i in range(n): # i is the index
    print('Index', i, 'Letter', string[i])

Index 0 Letter a
Index 1 Letter b
Index 2 Letter c
Index 3 Letter d
Index 4 Letter e


In general, `range` can have 3 parameters to pass: a start, a stop, and a step size. Let's see some examples.

In [48]:
# To get a list when using range, we need to cast it to a list
# Parameters: start, stop, step size
list(range(0, 101, 10))

[0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100]

In [49]:
# Default step size is 1
# Notice that 11 is not included
list(range(0, 11))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [50]:
# Default start is 0
list(range(6))

[0, 1, 2, 3, 4, 5]

The **enumerate** function is another useful function to use with `for` loops. It returns both the index and the item in each loop. 

In [51]:
string = 'abcde'

for i,letter in enumerate(string):
    print('Index', i,'Letter:', letter)

Index 0 Letter: a
Index 1 Letter: b
Index 2 Letter: c
Index 3 Letter: d
Index 4 Letter: e


<a id="2.2.3-while-loops"></a>

### 2.2.3 while Loops

The `while` statement in Python is another way to perform iteration. A `while` statement will repeatedly execute a single statement or group of statements as long as the condition is true. The reason it is called a 'loop' is because the code statements are looped through over and over again until the condition is no longer met.

The general format of a `while` loop is:

    while test:
        code to execute -> perform action 1
    else:
        code to execute -> perform action 2

Let’s look at a few simple <code>while</code> loops in action. 

In [52]:
x = 0

while x < 5:
    print('x is currently: ', x)
    print('x is still less than 5, adding 1 to x')
    x+=1

x is currently:  0
x is still less than 5, adding 1 to x
x is currently:  1
x is still less than 5, adding 1 to x
x is currently:  2
x is still less than 5, adding 1 to x
x is currently:  3
x is still less than 5, adding 1 to x
x is currently:  4
x is still less than 5, adding 1 to x


We can also add an <code>else</code> statement:

In [53]:
x = 0

while x < 5:
    print('x is currently: ',x)
    print(' x is still less than 5, adding 1 to x')
    x+=1
else:
    print('All Done!')

x is currently:  0
 x is still less than 5, adding 1 to x
x is currently:  1
 x is still less than 5, adding 1 to x
x is currently:  2
 x is still less than 5, adding 1 to x
x is currently:  3
 x is still less than 5, adding 1 to x
x is currently:  4
 x is still less than 5, adding 1 to x
All Done!


When working with `while` statements, it is important to ensure that the loop's termination condition will eventually be met. An infinite loop occurs when the condition controlling a while loop perpetually evaluates to True, preventing the loop from ever exiting.

Here is an example. 

In [54]:
## DO NOT RUN THIS CODE!!!! 
# while True:
#    print("I'm stuck in an infinite loop!")

If you did run the above cell, click on the Kernel menu above to interupt the kernel and restart the kernel!

<a id="2.2.4-break,-continue,-pass-statements"></a>

### 2.2.4 break, continue, pass Statements

We can use `break`, `continue`, and `pass` statements in our loops to add additional functionality for various cases. 

With the `break` and `continue` statements, the general format of the `while` loop looks like this:

    while test: 
        code to execute -> perform action 1
        if test: 
            break         # Exit the 'while' loop now
            continue      # Skip the Skip the statements after it, and go to top of the 'while' loop  
        else:
            code to execute -> perform action 2   # Run these statements when the 'if' test is False 

The `break` and `continue` statements can appear anywhere inside the loop’s body, but they are usually nested in an `if` statement to perform an action based on some condition.

In [55]:
for letter in "string":
    if letter == "i":
        break # exit the 'for' loop now
    print(letter)

print("The end")

s
t
r
The end


In [56]:
for letter in "string":
    if letter == "i":
        continue # go to the top of the 'for' loop now (skip the commands following 'continue')
    print(letter)

print("The end")

s
t
r
n
g
The end


Two more examples follow with an `else` statement.

In [57]:
x = 0

while x < 5:
    print('x is currently: ', x)
    print(' x is still less than 5, adding 1 to x')
    x += 1
    if x == 3:
        print('Breaking because x == 3')
        break  # terminate the 'while' loop, go to the 'print('The end')' statement
    else:
        print('continuing...')
        
print('The end')

x is currently:  0
 x is still less than 5, adding 1 to x
continuing...
x is currently:  1
 x is still less than 5, adding 1 to x
continuing...
x is currently:  2
 x is still less than 5, adding 1 to x
Breaking because x == 3
The end


In [58]:
x=0

while x < 5:
    print('x is currently: ', x)
    print(' x is still less than 5, adding 1 to x')
    x += 1
    if x == 3:
        print('Continuing to the next step')
        continue  # Skip the rest of the lines, and go to the while loop
        print('This line will be skipped and will not be printed')
    else:
        print('continuing...')
        
print('The end')

x is currently:  0
 x is still less than 5, adding 1 to x
continuing...
x is currently:  1
 x is still less than 5, adding 1 to x
continuing...
x is currently:  2
 x is still less than 5, adding 1 to x
Continuing to the next step
x is currently:  3
 x is still less than 5, adding 1 to x
continuing...
x is currently:  4
 x is still less than 5, adding 1 to x
continuing...
The end


The statement `pass` is generally used as a placeholder and it does not do anything. Suppose we have a loop or a function that is not implemented yet, but we want to implement it in the future. They cannot have an empty body, because this would give an error. So, we use the `pass` statement to construct a body that does nothing.

In [59]:
# Pass is just a placeholder for functionality to be added later
sequence = {'p', 'a', 's', 's'}
for val in sequence:
    pass

In [60]:
# Pass can be used as a placeholder for a function or a class
def my_function(arguments):
    pass

class Example:
    pass

<a id="appendix:-python-interpreter"></a>

## Appendix: Python Interpreter

**The material in the Appendix is not required for quizzes and assignments.**

The **interpreter** in Python is the program that executes other programs. When you run your programs in Python, the interpreter reads your programs, and carries out the instructions contained in the program. Or, we can say that the interpreter interprets your codes and enables the hardware on your computer to execute the program.

When you install Python on your computer, the Python interpreter will be part of the installation, either as an executable program, or as a set of linked libraries. Note that there are several different Python installations, and depending on the type of Python installation you have on your computer, the interpreter may be implemented as a C program, a set of Java classes, or
in another programming language. 

Understanding how the programs are executed in Python can be helpful for programmers. For instance, I saved the following simple file as *module1.py*.

<img style="float: left; height:180px;" src="images/pic2.jpg">

When I run the file in the Command Prompt, Python executed the file, and the output of the program is `Hello world!` and `10`.

<img style="float: left; height:240px;" src="images/pic3.jpg">

When we run programs in Python, the programs are first compiled into ***byte code***, and are afterward run by a  ***Python virtual machine (PVM)***, as shown in the figure below.

<img src="images/pic4.jpg" width="450">
<em>Figure source: Reference [1].</em>

**Byte code** is a format into which the **source code** (the statements in the file) is compiled by the Python interpreter. Byte-code is platform-independent (i.e., it can be run on Windows, Linux, MacOS), and it can be run more quickly than the source code program.  

The byte code is stored in a file with a *.pyc* extension, which stands for compiled .py file. The *.pyc* files are saved in a subdirectory named
\_\_pycache\_\_ located in the same directory where the source file is saved.

For example, the directory where *module1.py* is saved on my computer is shown below, and the \_\_pycache\_\_ subdirectory was automatically created by Python.  

<img style="float: left; height:180px;" src="images/pic5.jpg">

Within the subdirectory is the byte code file named *module1.cpython-36*. The name indicates that the Python installation on my computer uses the CPython interpreter, and the installed Python version is 3.6. Note that the file type is PYC file, meaning a .pyc extension.

<img style="float: left; height:140px;" src="images/pic6.jpg">

Byte code is saved for speed optimization. When I run module1.py next time, Python will skip the compilation step, and it will directly load the saved .pyc byte code file. However, if the original source code file module1.py was modified, Python will re-compile and update the byte code file. Similarly, if a different version of Python is installed, a new byte code file will be created that matches the current version of Python.

**Python virtual machine** (PVM) is the last part of the Python interpreter. PVM executes the byte code instructions one-by-one, i.e., it is the component that runs the programs. PVM is not a separate program, and it  does not need to be installed separately: it is part of the Python installation. PVM needs a software layer to allocate physical computing resources—such as processors, memory, and storage. 

Python belongs to the group of **interpreted languages**, or they are also called *scripting languages* (other languages in this group are Perl, Ruby, and JavaScript). As we explained, the Python interpreter reads the statements in source files and converts them into byte code files, which are afterwards executed by the PVM. Conversely, Java, C, and C++ belong to the group of **compiled languages**. In these languages, a compiler converts the statements in source files into binary machine code, which is afterwards executed by the computer hardware. Note that byte code files are different than binary machine code files. Consequently, running Python programs is slower than running C or C++ programs, because the code is interpreted as it is executed. On the other hand, writing and testing Python programs is faster and easier than writing and testing programs using compiled languages. (One last clarification: Python does compile source files, but the result is not a binary machine code, and because of that it is not considered a compiled language).

As we mentioned earlier, there are several different implementations of the Python interpreter. They include CPython, Jython, IronPython, Stackless Python, and PyPy. CPython is the standard, original implementation of Python, Jython is a Python implementation targeted for integration with the Java programming language, IronPython was designed to allow Python programs to integrate with applications coded to work with Microsoft’s .NET Framework for Windows, etc.

<a id="references"></a>

## References

1. Mark Lutz, "Learning Python," 5-th edition, O-Reilly, 2013. ISBN: 978-1-449-35573-9.
2. Pierian Data Inc., "Complete Python 3 Bootcamp," codes available at: [https://github.com/Pierian-Data/Complete-Python-3-Bootcamp](https://github.com/Pierian-Data/Complete-Python-3-Bootcamp).
3. Course T81 558:Applications of Deep Neural Networks, Washington University in St. Louis, Instructor: Jeff Heaton, codes available at: [https://github.com/jeffheaton/t81_558_deep_learning](https://github.com/jeffheaton/t81_558_deep_learning)

[BACK TO TOP](#top)